envoy/common/io/io_uring.h

```diff
@@ -0,0 +1,446 @@
+   * On read request completed.
+   * TODO (soulxu): wrap the raw result into a type. It can be `IoCallUint64Result`.
+   * @param req the ReadRequest object which is as request user data.
+   * @param result the result of operation in the request.
+   * @param injected indicates the completion is injected or not.
+   */
+  virtual void onRead(Request* req, int32_t result, bool injected) PURE;
+
+  /**
+   * On write request completed.
+   * TODO (soulxu): wrap the raw result into a type. It can be `IoCallUint64Result`.
+   * @param req the WriteRequest object which is as request user data.
+   * @param result the result of operation in the request.
+   * @param injected indicates the completion is injected or not.
+   */
+  virtual void onWrite(Request* req, int32_t result, bool injected) PURE;
+
+  /**
+   * On close request completed.
+   * TODO (soulxu): wrap the raw result into a type. It can be `IoCallUint64Result`.
+   * @param req the request object which is as request user data.
+   * @param result the result of operation in the request.
+   * @param injected indicates the completion is injected or not.
+   */
+  virtual void onClose(Request* req, int32_t result, bool injected) PURE;
+
+  /**
+   * On cancel request completed.
+   * TODO (soulxu): wrap the raw result into a type. It can be `IoCallUint64Result`.
+   * @param req the request object which is as request user data.
+   * @param result the result of operation in the request.
+   * @param injected indicates the completion is injected or not.
+   */
+  virtual void onCancel(Request* req, int32_t result, bool injected) PURE;
+
+  /**
+   * On shutdown request completed.
+   * TODO (soulxu): wrap the raw result into a type. It can be `IoCallUint64Result`.
+   * @param req the request object which is as request user data.
+   * @param result the result of operation in the request.
+   * @param injected indicates the completion is injected or not.
+   */
+  virtual void onShutdown(Request* req, int32_t result, bool injected) PURE;
+
+  /**
+   * Inject a request completion to the io uring instance.
+   * @param type the request type of injected completion.
+   */
+  virtual void injectCompletion(uint32_t type) PURE;
+
+  /**
+   * Return the current status of IoUringSocket.
+   * @return the status.
+   */
+  virtual IoUringSocketStatus getStatus() const PURE;
+
+  virtual const OptRef<ReadParam>& getReadParam() const PURE;
+  virtual const OptRef<AcceptedSocketParam>& getAcceptedSocketParam() const PURE;
+  virtual const OptRef<WriteParam>& getWriteParam() const PURE;
+
+  virtual void clearAcceptedSocketParam() PURE;
+  virtual void setFileReadyCb(Event::FileReadyCb cb) PURE;
+};
+
+/**
+ * Abstract for per-thread worker.
+ */
+class IoUringWorker : public ThreadLocal::ThreadLocalObject {
+public:
+  ~IoUringWorker() override = default;
+
+  /**
+   * Add an accept socket socket to the worker.
+   */
+  virtual IoUringSocket& addAcceptSocket(os_fd_t fd, Event::FileReadyCb cb,
```

Comment:
It is unclear how the caller is going to use the **reference** to the IoUringSocket. The IoUringSocket interface is also confusing. It has both methods for submitting I/O operations and callbacks for completions. How are the callback are going to be used?

Reply:
The IoUringSocket manages the lifetime of fd, the caller(IoUringSocketHandleImpl ) can use it to register a callback to listen the read/write event. Also can use it to submit write request, or activate a read/write event(through injectCompletion()).

But IoUringSocket is owned by the IoUringWorker, so the IoUringSocketHandleImpl only got a reference to it. And IoUringSocket may have a longer life time than the IoUringSocketHandleImpl (like the close of socket is async also).

The two pictures in this doc explain the responsibility for all the components and how they interact more
https://docs.google.com/document/d/17M15J6vh7ZzoJPCY65j6HNoLSqNEoWnQ3Oa9oKIru0o/edit#heading=h.21fg36mquy3o

Let me know if there still has something confusing, I can explain more.